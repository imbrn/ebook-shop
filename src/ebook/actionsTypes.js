export const FETCH_ALL_EBOOKS = "FETCH_ALL_EBOOKS";
