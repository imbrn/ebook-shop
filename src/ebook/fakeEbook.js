import Ebook from "./Ebook";

export function fakeEbook(id) {
  return new Ebook({
    id,
    name: `Ebook ${id}`,
    description: `The ebook ${id}`,
    price: id * 10000
  });
}

export function fakeEbooks(amount = 10, firstId = 1) {
  return Array(amount)
    .fill()
    .map(() => fakeEbook(firstId++));
}
