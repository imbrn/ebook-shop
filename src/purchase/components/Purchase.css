.root {
  padding: 0 20px;

  & > header,
  & > footer {
    padding: 40px 0;
  }

  & > footer {
    display: flex;
    justify-content: space-between;
  }
}

.about {
  display: flex;
  flex-direction: column;
}

.flags {
  svg {
    width: auto;
    height: 26px;
    margin: 0 2px;
  }
}
