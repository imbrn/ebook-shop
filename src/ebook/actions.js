import { getService } from "../service";
import { FETCH_ALL_EBOOKS } from "./actionsTypes";

export function fetchAllEbooks() {
  return dispatch => {
    dispatch(onFetchAllEbooks("request"));
    return getService()
      .fetchAllEbooks()
      .then(ebooks => dispatch(onFetchAllEbooks("success", { ebooks })));
  };
}

function onFetchAllEbooks(status, props) {
  return {
    type: FETCH_ALL_EBOOKS,
    status,
    ...props
  };
}
