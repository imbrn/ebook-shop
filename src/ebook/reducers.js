import { FETCH_ALL_EBOOKS } from "./actionsTypes";

export function ebooks(state = [], action) {
  switch (action.type) {
    case FETCH_ALL_EBOOKS:
      return action.ebooks;
    default:
      return state;
  }
}
