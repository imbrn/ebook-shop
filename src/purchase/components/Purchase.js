import React, { Fragment } from "react";
import { Text, Separator, Circle } from "../../common/components";
import { connect } from "react-redux";
import EbooksList from "./EbooksList";
import DataForm from "./DataForm";
import styles from "./Purchase.css";
import BarCodeSvg from "../../common/icons/bar-code.svg";
import MasterCardSvg from "../../common/icons/mastercard.svg";
import VisaSvg from "../../common/icons/visa.svg";
import SuccessSvg from "../../common/icons/success.svg";

export const Purchase = ({ ebooks, purchase }) => {
  const renderContent = () => {
    switch (ebooks.status.kind) {
      case "request":
        return renderContentLoading();
      case "fail":
        return renderContentFail();
      case "success":
        return renderContentLoaded();
      default:
        return null;
    }
  };

  const renderContentLoading = () => <div>Loading...</div>;
  const renderContentFail = () => <div>Fail</div>;

  const renderContentLoaded = () => {
    switch (purchase.status.kind) {
      case "success":
        return <PurchaseSuccess purchase={purchase} />;
      case "fail":
        return renderPurchaseFail();
      case "initial":
        return renderPurchasing();
      default:
        return null;
    }
  };

  const renderPurchaseFail = () => <div>Purchase failed</div>;

  const renderPurchasing = () => (
    <Fragment>
      <Text>
        Welcome to the best place for you to learn about Latin América
        E-commerce. Start to learn now and discovery ways options to improve
        your sales.
      </Text>
      <EbooksList />
      <DataForm />
    </Fragment>
  );

  return (
    <div className={styles.root}>
      <header>
        <Text bold size="medium">
          E-book Store
        </Text>
      </header>

      <div>{renderContent()}</div>

      <Separator />

      <footer>
        <div className={styles.about}>
          <Text bold>E-book Store</Text>
          <Text>Powered by EBANX • Products B2B</Text>
        </div>

        <div className={styles.flags}>
          <BarCodeSvg />
          <VisaSvg />
          <MasterCardSvg />
        </div>
      </footer>
    </div>
  );
};

const PurchaseSuccess = ({ purchase }) => {
  const renderMessage = () => {
    if (purchase.status.data.payment.method === "boleto") {
      return (
        <Text>
          The boleto was created with success and sent to email{" "}
          <Text bold>{purchase.status.data.personalData.email}</Text>
        </Text>
      );
    } else {
      return (
        <Text>
          The payment using the credit card{" "}
          <Text bold>
            {hideSomeDigits(purchase.status.data.payment.cardNumber)}
          </Text>{" "}
          was made successfully
        </Text>
      );
    }
  };

  return (
    <div className={styles.purchaseSuccess}>
      <div className={styles.successMessage}>
        <Circle className={styles.successMessageCircle}>
          <SuccessSvg />
        </Circle>
        <Text bold>Purchase realized with success!</Text>
      </div>
      <div>{renderMessage()}</div>
    </div>
  );
};

function hideSomeDigits(number) {
  return number.split("").map((char, index) => {
    if (index > 3 && index < 12) return "*";
    return char;
  });
}

const mapStateToProps = state => ({
  ebooks: state.ebooks,
  purchase: state.purchase
});

export default connect(mapStateToProps)(Purchase);
