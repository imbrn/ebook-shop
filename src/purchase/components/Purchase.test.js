import React from "react";
import { render, cleanup } from "react-testing-library";
import { Purchase } from "./Purchase";
import { fakeEbooks } from "../../ebook/fakeEbook";

jest.mock("nanoid", () => {
  let nextId = 1;
  return () => nextId++;
});

afterEach(cleanup);

describe("renders correctly", () => {
  test("initial", () => {
    const purchase = {
      ebooks: fakeEbooks(2),
      personalData: {
        name: "Customer Name",
        email: "customer@email.com",
        cpf: "123456789"
      },
      billingAddress: {
        zipCode: "12345678",
        state: "My State",
        city: "My City",
        address: "My Address, 123"
      },
      payment: {
        method: "creditCard",
        cardholderName: "Customer Name",
        cardNumber: "1234567891011121",
        dueDate: "12/25",
        cvv: "123"
      },
      status: {
        type: "initial"
      }
    };

    const { container } = render(<Purchase purchase={purchase} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("purchased with success", () => {
    const purchase = {
      ebooks: fakeEbooks(2),
      personalData: {
        name: "Customer Name",
        email: "customer@email.com",
        cpf: "123456789"
      },
      billingAddress: {
        zipCode: "12345678",
        state: "My State",
        city: "My City",
        address: "My Address, 123"
      },
      payment: {
        method: "creditCard",
        cardholderName: "Customer Name",
        cardNumber: "1234567891011121",
        dueDate: "12/25",
        cvv: "123"
      },
      status: {
        type: "success",
        purchaseId: 123
      }
    };

    const { container } = render(<Purchase purchase={purchase} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
