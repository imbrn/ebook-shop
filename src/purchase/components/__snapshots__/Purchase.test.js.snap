// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly fetching ebooks 1`] = `
.c0 {
  font-weight: bold;
  font-size: 18px;
}

.c2 {
  font-weight: bold;
}

.c1 {
  border: none;
  height: 2px;
  background: #dedede;
}

<div>
  <header>
    <span
      class="c0"
    >
      E-book Store
    </span>
  </header>
  <div>
    <div>
      Loading...
    </div>
  </div>
  <hr
    class="c1"
  />
  <footer>
    <div>
      <span
        class="c2"
      >
        E-book Store
      </span>
      <span
        class=""
      >
        Powered by EBANX • Products B2B
      </span>
    </div>
    <div />
  </footer>
</div>
`;

exports[`renders correctly fetching ebooks failed 1`] = `
.c0 {
  font-weight: bold;
  font-size: 18px;
}

.c2 {
  font-weight: bold;
}

.c1 {
  border: none;
  height: 2px;
  background: #dedede;
}

<div>
  <header>
    <span
      class="c0"
    >
      E-book Store
    </span>
  </header>
  <div>
    <div>
      Fail
    </div>
  </div>
  <hr
    class="c1"
  />
  <footer>
    <div>
      <span
        class="c2"
      >
        E-book Store
      </span>
      <span
        class=""
      >
        Powered by EBANX • Products B2B
      </span>
    </div>
    <div />
  </footer>
</div>
`;

exports[`renders correctly purchasing fail 1`] = `
.c0 {
  font-weight: bold;
  font-size: 18px;
}

.c2 {
  font-weight: bold;
}

.c1 {
  border: none;
  height: 2px;
  background: #dedede;
}

<div>
  <header>
    <span
      class="c0"
    >
      E-book Store
    </span>
  </header>
  <div>
    <div>
      Purchase failed
    </div>
  </div>
  <hr
    class="c1"
  />
  <footer>
    <div>
      <span
        class="c2"
      >
        E-book Store
      </span>
      <span
        class=""
      >
        Powered by EBANX • Products B2B
      </span>
    </div>
    <div />
  </footer>
</div>
`;

exports[`renders correctly purchasing initial 1`] = `
.c1 {
  position: relative;
  width: 20px;
  height: 20px;
  border: 2px solid black;
}

.c3 {
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.c2 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c0 {
  font-weight: bold;
  font-size: 18px;
}

.c4 {
  font-weight: bold;
}

.c5 {
  border: none;
  height: 2px;
  background: #dedede;
}

.c6 {
  display: inline-block;
  width: 36px;
  height: 36px;
  text-align: center;
  line-height: 32px;
  border-radius: 100%;
  font-weight: bold;
  border: 2px solid black;
  margin-right: 10px;
}

<div>
  <header>
    <span
      class="c0"
    >
      E-book Store
    </span>
  </header>
  <div>
    <span
      class=""
    >
      Welcome to the best place for you to learn about Latin América E-commerce. Start to learn now and discovery ways options to improve your sales.
    </span>
    <div>
      <ul>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="1"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/1.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 1
              </span>
              <span
                class=""
              >
                The ebook 1
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 100,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="2"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/2.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 2
              </span>
              <span
                class=""
              >
                The ebook 2
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 200,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="3"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/3.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 3
              </span>
              <span
                class=""
              >
                The ebook 3
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 300,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="4"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/4.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 4
              </span>
              <span
                class=""
              >
                The ebook 4
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 400,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="5"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/5.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 5
              </span>
              <span
                class=""
              >
                The ebook 5
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 500,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="6"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/6.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 6
              </span>
              <span
                class=""
              >
                The ebook 6
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 600,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="7"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/7.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 7
              </span>
              <span
                class=""
              >
                The ebook 7
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 700,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="8"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/8.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 8
              </span>
              <span
                class=""
              >
                The ebook 8
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 800,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="9"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/9.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 9
              </span>
              <span
                class=""
              >
                The ebook 9
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 900,00
            </span>
          </div>
        </li>
        <li>
          <div>
            <div
              class="c1"
            >
              <div
                class="c2"
              />
              <input
                class="c3"
                type="checkbox"
                value="10"
              />
            </div>
            <img
              class=""
              src="public/images/ebooks/10.png"
            />
            <div>
              <span
                class="c4"
              >
                Ebook 10
              </span>
              <span
                class=""
              >
                The ebook 10
              </span>
            </div>
            <span
              class="c0"
            >
              R$ 1000,00
            </span>
          </div>
        </li>
      </ul>
      <hr
        class="c5"
      />
      <div>
        <span
          class="c0"
        >
          TOTAL
        </span>
        <span
          class="c0"
        >
          R$ 0,00
        </span>
      </div>
    </div>
    <form>
      <div
        class=""
      >
        <header>
          <span
            class="c6"
          >
            1
          </span>
          <span
            class="c4"
          >
            Personal data
          </span>
        </header>
      </div>
      <div
        class=""
      >
        <header>
          <span
            class="c6"
          >
            2
          </span>
          <span
            class="c4"
          >
            Billing Address
          </span>
        </header>
      </div>
      <div
        class=""
      >
        <header>
          <span
            class="c6"
          >
            3
          </span>
          <span
            class="c4"
          >
            Payment
          </span>
        </header>
      </div>
    </form>
  </div>
  <hr
    class="c5"
  />
  <footer>
    <div>
      <span
        class="c4"
      >
        E-book Store
      </span>
      <span
        class=""
      >
        Powered by EBANX • Products B2B
      </span>
    </div>
    <div />
  </footer>
</div>
`;

exports[`renders correctly purchasing success 1`] = `
.c0 {
  font-weight: bold;
  font-size: 18px;
}

.c2 {
  font-weight: bold;
}

.c1 {
  border: none;
  height: 2px;
  background: #dedede;
}

<div>
  <header>
    <span
      class="c0"
    >
      E-book Store
    </span>
  </header>
  <div>
    <div>
      Purchase success
    </div>
  </div>
  <hr
    class="c1"
  />
  <footer>
    <div>
      <span
        class="c2"
      >
        E-book Store
      </span>
      <span
        class=""
      >
        Powered by EBANX • Products B2B
      </span>
    </div>
    <div />
  </footer>
</div>
`;
